File: app/src/main/java/net/mbonnin/arcanetracker/CardsAdapter.java
package net.mbonnin.arcanetracker;

import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.graphics.drawable.BitmapDrawable;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.squareup.picasso.Callback;
import com.squareup.picasso.NetworkPolicy;
import com.squareup.picasso.Picasso;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Locale;

import timber.log.Timber;

/**
 * Created by martin on 10/21/16.
 */
public class CardsAdapter extends RecyclerView.Adapter {
    private final CardDb cardDb;
    private int mClassIndex;
    private ArrayList<Card> mCardList = new ArrayList<>();
    private Listener mListener;
    private String mSearchQuery;
    private int mCost = -1;
    private ArrayList<String> mDisabledCards = new ArrayList<>();

    public void setCost(int cost) {
        mCost = cost;
        filter();
    }

    public CardsAdapter(CardDb cardDb) {
        this.cardDb = cardDb;
    }

    public void setDisabledCards(ArrayList<String> list) {
        mDisabledCards = list;
        notifyDataSetChanged();
    }

    public void setListener(Listener listener) {
        mListener = listener;
    }

    public interface Listener {
        void onClick(Card card);
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.card_view, null);

        RecyclerView.ViewHolder holder = new RecyclerView.ViewHolder(view) {
        };

        ImageView imageView = (ImageView) view.findViewById(R.id.imageView);
        ((AspectRatioImageView)imageView).setAspectRatio(1.51f);

        view.setOnTouchListener((v, event) -> {
            if (event.getActionMasked() == MotionEvent.ACTION_DOWN) {
                if (mDisabledCards.contains(mCardList.get(holder.getAdapterPosition()).id)) {
                    return false;
                }
                imageView.setColorFilter(Color.argb(120, 255, 255, 255), PorterDuff.Mode.SRC_OVER);
                return true;
            } else if (event.getActionMasked() == MotionEvent.ACTION_CANCEL || event.getActionMasked() == MotionEvent.ACTION_UP) {
                imageView.clearColorFilter();

                if (event.getActionMasked() == MotionEvent.ACTION_UP) {
                    int position = holder.getAdapterPosition();
                    // not really sure how we could go outside this condition but it happens...
                    if (position < mCardList.size()) {
                        mListener.onClick(mCardList.get(position));
                    }
                }
            }
            return false;
        });
        return holder;
    }

    public void setClass(int classIndex) {
        mClassIndex = classIndex;

        filter();
    }

    public void setSearchQuery(String searchQuery) {
        if (searchQuery.equals("")) {
            mSearchQuery = null;
        } else {
            mSearchQuery = searchQuery.toLowerCase();
        }
        filter();

    }

    private void filter() {
        mCardList.clear();
        List<Card> allCards = cardDb.getCards();

        String playerClass = Card.classIndexToPlayerClass(mClassIndex);
        for (Card card : allCards) {
            if (card.collectible == null || !card.collectible) {
                continue;
            }

            if (card.cost == null) {
                continue;
            }

            if (mCost != -1) {
                if (mCost == 7) {
                    if (card.cost < 7) {
                        continue;
                    }
                } else if (card.cost != mCost) {
                    continue;
                }
            }

            /*if (card.type == null
                    || (!card.type.equals("SPELL") && !card.type.equals("MINION") && !card.type.equals("WEAPON"))) {
                continue;
            }*/
            if (card.playerClass != null && playerClass == null
                    || (card.playerClass == null && playerClass != null)
                    || (playerClass != null && !playerClass.equals(card.playerClass))) {
                continue;
            }

            if (mSearchQuery != null) {
                boolean found = false;
                if (card.text != null && card.text.toLowerCase().indexOf(mSearchQuery) != -1) {
                    found = true;
                }

                if (!found && card.name != null && card.name.toLowerCase().indexOf(mSearchQuery) != -1) {
                    found = true;
                }

                if (!found && card.race != null && card.race.toLowerCase().indexOf(mSearchQuery) != -1) {
                    found = true;
                }

                if (!found) {
                    continue;
                }
            }

            mCardList.add(card);
        }

        Collections.sort(mCardList, (a, b) -> a.cost - b.cost);

        notifyDataSetChanged();
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        ImageView imageView = (ImageView) holder.itemView.findViewById(R.id.imageView);
        TextView textView = (TextView) holder.itemView.findViewById(R.id.textView);

        Card card = mCardList.get(position);

        if (mDisabledCards.contains(card.id)) {
            imageView.setColorFilter(Color.argb(180, 0, 0, 0), PorterDuff.Mode.SRC_ATOP);
        } else {
            imageView.clearColorFilter();
        }

        String url = "card://" + card.id;
        Timber.d("fetching " + url);
        int placeHolderRes;
        if (card.rarity.equals(Card.RARITY_LEGENDARY)) {
            placeHolderRes = R.raw.placeholder_legendary;
        } else if (card.type.equals(Card.TYPE_SPELL)) {
            placeHolderRes = R.raw.placeholder_spell;
        } else {
            placeHolderRes = R.raw.placeholder_minion;
        }

        textView.setText(card.name);
        Picasso.with(imageView.getContext())
                .load(url)
                .placeholder(placeHolderRes)
                .into(imageView, new Callback() {
                    @Override
                    public void onSuccess() {
                        textView.setText("");
                    }

                    @Override
                    public void onError() {

                    }
                });
    }

    @Override
    public int getItemCount() {
        return mCardList.size();
    }
}

File: app/src/main/java/net/mbonnin/arcanetracker/CardsAdapter.kt
package net.mbonnin.arcanetracker

import android.graphics.Color
import android.graphics.PorterDuff
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import com.squareup.picasso.Callback
import com.squareup.picasso.Picasso
import net.mbonnin.hsmodel.Card
import net.mbonnin.hsmodel.CardJson
import java.util.ArrayList
import kotlin.Comparator

/**
 * Created by martin on 10/21/16.
 */
class CardsAdapter : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    private var mClass: String? = null
    private val mCardList = ArrayList<Card>()
    private var mListener: Listener? = null
    private var mSearchQuery: String? = null
    private var mCost = -1
    private var mDisabledCards = ArrayList<String>()

    fun setCost(cost: Int) {
        mCost = cost
        filter()
    }

    fun setDisabledCards(list: ArrayList<String>) {
        mDisabledCards = list
        notifyDataSetChanged()
    }

    fun setListener(listener: Listener) {
        mListener = listener
    }

    interface Listener {
        fun onClick(card: Card)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.card_view, null)

        val holder = object : RecyclerView.ViewHolder(view) {

        }

        val imageView = view.findViewById<ImageView>(R.id.imageView)
        (imageView as AspectRatioImageView).setAspectRatio(1.51f)

        view.setOnTouchListener { v, event ->
            val position = holder.adapterPosition
            /**
             * the NO_POSITION case could happen if you click very fast.
             * not sure about the other case..., maybe it's not needed anymore
             */
            if (position == RecyclerView.NO_POSITION || position >= mCardList.size) {
                return@setOnTouchListener false
            }

            if (event.actionMasked == MotionEvent.ACTION_DOWN) {
                if (mDisabledCards.contains(mCardList[position].id)) {
                    return@setOnTouchListener false
                }
                imageView.setColorFilter(Color.argb(120, 255, 255, 255), PorterDuff.Mode.SRC_OVER)
                return@setOnTouchListener true
            } else if (event.actionMasked == MotionEvent.ACTION_CANCEL || event.actionMasked == MotionEvent.ACTION_UP) {
                imageView.clearColorFilter()

                if (event.actionMasked == MotionEvent.ACTION_UP) {
                    mListener!!.onClick(mCardList[position])
                }
            }
            false
        }
        return holder
    }

    fun setClass(clazz: String) {
        mClass = clazz

        filter()
    }

    fun setSearchQuery(searchQuery: String) {
        if (searchQuery == "") {
            mSearchQuery = null
        } else {
            mSearchQuery = searchQuery.toLowerCase()
        }
        filter()

    }

    private fun filter() {
        mCardList.clear()
        val allCards = CardJson.allCards()

        for (card in allCards) {
            if (!card.collectible) {
                continue
            }

            if (card.cost == null) {
                continue
            }

            if (mCost != -1) {
                if (mCost == 7) {
                    if (card.cost ?: 0 < 7) {
                        continue
                    }
                } else if (card.cost != mCost) {
                    continue
                }
            }

            if (mClass != card.playerClass) {
                continue
            }

            if (mSearchQuery != null) {
                var found = false
                if (card.text != null && card.text!!.toLowerCase().contains(mSearchQuery!!)) {
                    found = true
                }

                if (!found && card.name.toLowerCase().contains(mSearchQuery!!)) {
                    found = true
                }

                if (!found && card.race != null && card.race!!.toLowerCase().contains(mSearchQuery!!)) {
                    found = true
                }

                if (!found) {
                    continue
                }
            }

            mCardList.add(card)
        }

        mCardList.sortWith(Comparator{ a, b ->
            val r =  Utils.compareNullSafe(a.cost, b.cost)
            if (r != 0) {
                return@Comparator r
            }
            a.name.compareTo(a.name)
        })

        notifyDataSetChanged()
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val imageView = holder.itemView.findViewById<ImageView>(R.id.imageView)
        val textView = holder.itemView.findViewById<TextView>(R.id.textView)

        val (id, _, _, _, _, name) = mCardList[position]

        if (mDisabledCards.contains(id)) {
            imageView.setColorFilter(Color.argb(180, 0, 0, 0), PorterDuff.Mode.SRC_ATOP)
        } else {
            imageView.clearColorFilter()
        }

        val context = textView.context

        textView.text = name

        Picasso.with(context).load(Utils.getCardUrl(id)).into(imageView, object : Callback {
            override fun onSuccess() {
                textView.text = ""
            }

            override fun onError() {}
        })
    }

    override fun getItemCount(): Int {
        return mCardList.size
    }
}
